db/db_impl/db_impl_open.cc

```diff
@@ -582,21 +582,40 @@ Status DBImpl::Recover(
               "existing log file: ",
               file);
         } else {
-          logs.push_back(number);
+          wal_files[number] =
+              LogFileName(immutable_db_options_.wal_dir, number);
         }
       }
     }
 
-    if (logs.size() > 0) {
+    if (immutable_db_options_.track_and_verify_wals_in_manifest) {
+      // Verify WALs in MANIFEST.
+      s = versions_->GetWalSet().CheckWals(env_, wal_files);
+    } else if (!versions_->GetWalSet().GetWals().empty()) {
+      // Tracking is disabled, clear previously tracked WALs from MANIFEST,
+      // otherwise, in the future, if WAL tracking is enabled again,
+      // since the WALs deleted when WAL tracking is disabled are not persisted
+      // into MANIFEST, WAL check may fail.
+      VersionEdit edit;
+      for (const auto& wal : versions_->GetWalSet().GetWals()) {
+        WalNumber number = wal.first;
+        edit.DeleteWal(number);
```

Comment:
In `WalSet`, it's possible to mark wals with smaller file number as deleted, as shown in `WalSet::DeleteWalsBefore()`.
When we write MANIFEST, can we log the operation as "mark all wals smaller than a specific file number as deleted"?

Reply:
this is what I'm thinking of doing, it depends on how we are going to write the deleted WALs into MANIFEST. If we write the deleted WAL edits in PurgeObsoleteFiles, then we need to delete individual WALs. If we piggyback the deleted WAL edits during flushing, then we can use `DeleteWalsBefore`. 
I prefer to making this decision in next PR where we decide where to delete the WALs. In my current prototype, I'm going to use `DeleteWalsBefore`, so the `DeleteWal` API might be removed in next PR.